runtime/compiler/x/codegen/X86PrivateLinkage.cpp

```diff
@@ -1297,120 +1297,153 @@ void TR::X86CallSite::setupVirtualGuardInfo()
-                  }
+               topValue = 0;
                }
             }
+         }
 
-         if (!topValue && !callNode->getSymbolReference()->isUnresolved() &&
+      if (!topValue)
+         {
+         if (isVirtual &&
              (callNode->getSymbol()->castToMethodSymbol()->getRecognizedMethod() == TR::java_lang_Object_clone))
-            topValue = (uintptrj_t) comp()->getObjectClassPointer();
-
-         if (topValue)
             {
+            TR_ResolvedMethod *profiledVirtualMethod =
+                  methodSymRef->getOwningMethod(comp())->getResolvedVirtualMethod(comp(),
+                     (TR_OpaqueClassBlock *)comp()->getObjectClassPointer(),
+                     methodSymRef->getOffset());
 
-            TR_ResolvedMethod *profiledVirtualMethod = callNode->getSymbolReference()->getOwningMethod(comp())->getResolvedVirtualMethod(comp(),
-               (TR_OpaqueClassBlock *)topValue, methodSymRef->getOffset());
-            if (profiledVirtualMethod &&
-                (!profiledVirtualMethod->isInterpreted() ||
-                profiledVirtualMethod->isJITInternalNative()))
-               {
-               //if (!getMethodSymbol()->isInterface() && profiledVirtualMethod->isJITInternalNative())
-               //printf("New opportunity in %s to callee %s\n", comp()->signature(), profiledVirtualMethod->signature(comp()->trMemory(), stackAlloc));
-               //TR_ASSERT(profiledVirtualMethod->classOfMethod() == (TR_OpaqueClassBlock *)topValue, "assertion failure");
-
-               TR_OpaqueMethodBlock *methodToBeCompared = NULL;
-               int32_t slot = -1;
-               if (profiledVirtualMethod->isJITInternalNative())
-                  {
-                  int32_t offset = callNode->getSymbolReference()->getOffset();
-                  slot = fej9->virtualCallOffsetToVTableSlot(offset);
-                  methodToBeCompared = profiledVirtualMethod->getPersistentIdentifier();
-                  }
-
-               _profiledTargets->add(new(comp()->trStackMemory()) TR::X86PICSlot((uintptrj_t)topValue, profiledVirtualMethod, true, methodToBeCompared, slot));
-               }
+            tryToAddStaticPIC((uintptrj_t)comp()->getObjectClassPointer(), callNode, profiledVirtualMethod, fej9, 0, 0.0, true, true);
             }
          }
-      }
-   else if (getMethodSymbol()->isInterface())
-      {
-      bool staticPICsExist = false;
-      int32_t numStaticPICSlots = 0;
-
-
-      TR_AddressInfo *addressInfo = static_cast<TR_AddressInfo*>(TR_ValueProfileInfoManager::getProfiledValueInfo(callNode, comp(), AddressInfo));
-#if defined(OSX)
-      uint64_t topValue;
-#else
-      uintptr_t topValue;
-#endif /* OSX */
-      float missRatio = 0.0;
-      if (addressInfo && addressInfo->getTopValue(topValue) > 0 && topValue && !comp()->getPersistentInfo()->isObsoleteClass((void*)topValue, fej9) &&
-          addressInfo->getTopProbability() >= getMinProfiledCallFrequency())
+      else
```

Comment:
The following path was only originally taken for interface calls.  Now it seems that when a top value has been determined to be non-zero it will be taken for virtual calls as well, in which case the code looks considerably different.  Am I misreading the code or logic?

Reply:
I'll wait for clarification on this before I go further with this review as it will affect how I perceive the code.  I think I see what you're doing for virtuals because you have that escape hatch at the bottom of the loop that only allows you to add one PIC slot for virtuals (and I'm not sure yet I like that approach), but the logic to grab the top value looks quite a bit different.

Reply:
In the previous code, we would also add a static PIC for a virtual call if the top value was non-NULL. 

It was a bit complicated because in the virtual call case, even if the top value was NULL, it would add a PIC under special circumstances. That's why I have the check in https://github.com/eclipse/openj9/pull/8474/files#diff-bc6d5bca24d2675b48424ae7ce17bb9cR1423 , which basically only does the special thing if top value is NULL && the call is virtual.

In the case where top value is NOT NULL, then we add a single vPIC (to match existing behaviour), which is done by that approach you don't quite like :).